File: RK4Solver/flock.py
import numpy as np
from bird import Bird
import copy

class Flock():

    def __init__(self,
                 N = 10,
                 h = 0.001,
                 t = 0.0,
                 bird_starts=None,
                 LIA = False
                 ):
        self.h = h
        self.t = t
        self.N = N
        self.num_agents = N
        self.LIA = LIA

        self.total_vortices = 0.0
        self.total_dist = 0.0

        self.max_r = 1.0
        self.max_steps = 100.0/self.h

        self.energy_punishment = 2.0
        self.forward_reward = 5.0
        self.crash_reward = -100.0

        self.agents = range(self.N)
        birds = None
        if bird_starts is None:
            bird_starts = [dict(z = 50.0, y = 3.0*i, u=5.0) for i in range(self.N)]

        birds = [Bird(**start_val) for start_val in bird_starts]
        self.starting_conditions = [copy.deepcopy(bird) for bird in birds]
        self.birds = birds

    def reset(self):
        self.old_birds = [copy.deepcopy(self.birds[i]) for i in range(self.N)]
        birds = [copy.deepcopy(bird) for bird in self.starting_conditions]
        self.birds = birds

    def update_bird(self, action, agent):
        thrust = action[0]
        bird = self.birds[agent]
        self.update_angles(action, agent)
        vortices = self.get_vortices(bird)
        bird.update(thrust, self.h, vortices)

    def get_reward(self, action, agent):
        reward = 0
        done = False
        bird = self.birds[agent]
        if bird.x > bird.X[-2]:
            reward += self.forward_reward
        reward -= self.energy_punishment * action[0]

        if self.crashed(bird):
            done = True
            reward = self.crash_reward

        if bird.x > 500.0:
            done = True

        return done, reward

    def crashed(self, bird):
        if bird.z <= 0 or bird.z >= 100:
            return True

        lim = 2*np.pi
        if abs(bird.p) > lim or abs(bird.q) > lim or  abs(bird.r) > lim:
            return True

        crash = False
        for b in range(len(self.birds)):
            other = self.birds[b]
            if other is not bird:
                dist = np.sqrt((bird.x - other.x)**2 + (bird.y - other.y)**2 + (bird.z - other.z)**2)
                if dist < bird.Xl/2.0:
                    crash = True
        return crash

    def get_vortices(self, curr):
        vortices = []
        for b in range(len(self.birds)):
            bird = self.birds[b]
            if bird is not curr:
                for vorts in [bird.VORTICES_LEFT, bird.VORTICES_RIGHT]:
                    i = 0
                    v = vorts[i]

                    #want first vortex ahead of current vortex
                    while i < len(vorts) and v.x < curr.x:
                        v = vorts[i]
                        i = i+1
                    if i < len(vorts) and v.x >= curr.x:
                        r = np.sqrt((curr.y - v.y)**2 + (curr.z - v.z)**2)
                        if r < self.max_r:
                            vortices.append(v)
        return vortices

    def get_observation(self, agent):
        force = self.birds[agent].F
        torque = self.birds[agent].T

        bird = self.birds[agent]
        pos = []
        pos += [bird.x, bird.y, bird.z]
        pos += [bird.u, bird.v, bird.w]
        pos += [bird.p, bird.q, bird.r]
        pos += [bird.phi, bird.theta, bird.psi]
        pos += [bird.alpha_l, bird.beta_l, bird.alpha_r, bird.beta_r]
        nearest = bird.seven_nearest(self.birds)
        for other in nearest:
            pos += [other.x - bird.x, other.y - bird.y, other.z - bird.z]
            pos += [other.u - bird.u, other.v - bird.v, other.w - bird.w]

        obs = np.array(force + torque + pos)
        return obs

    def update_vortices(self):
        for b in range(len(self.birds)):
            bird = self.birds[b]
            bird.shed_vortices()

            if self.LIA:
                bird.update_vortex_positions(bird.VORTICES_RIGHT, self.h*10.0)
                bird.update_vortex_positions(bird.VORTICES_LEFT, self.h*10.0)

            #remove expired vortices
            if len(bird.VORTICES_LEFT) > 1.0/self.h:
                a = bird.VORTICES_LEFT.pop(0)
                b = bird.VORTICES_RIGHT.pop(0)

    def update_angles(self, action, agent):
        bird = self.birds[agent]
        limit_alpha = np.pi/6.0
        limit_beta = np.pi/4.0
        new_al = bird.alpha_l + action[1]
        if new_al > limit_alpha:
            new_al = limit_alpha
        if new_al < -limit_alpha:
            new_al = -limit_alpha
        bird.alpha_l = new_al

        new_bl = bird.beta_l + action[2]
        if new_bl > limit_beta:
            new_bl = limit_beta
        if new_bl < -limit_beta:
            new_bl = -limit_beta
        bird.beta_l = new_bl

        new_ar = bird.alpha_r + action[3]
        if new_ar > limit_alpha:
            new_ar = limit_alpha
        if new_ar < -limit_alpha:
            new_ar = -limit_alpha
        bird.alpha_r = new_ar

        new_br = bird.beta_r + action[4]
        if new_br > limit_beta:
            new_br = limit_beta
        if new_br < -limit_beta:
            new_br = -limit_beta
        bird.beta_r = new_br


    def get_birds(self):
        return self.birds


    def print_bird(self, bird, action = []):
        print('-----------------------------------------------------------------------')
        print(self.agent_selection)
        print("thrust, al, ar, bl, br \n", action)
        print("x, y, z: \t\t", [bird.x, bird.y, bird.z])
        print("u, v, w: \t\t", [bird.u, bird.v, bird.w])
        print("phi, theta, psi: \t", [bird.phi, bird.theta, bird.psi])
        print("p, q, r: \t\t", [bird.p, bird.q, bird.r])
        print("alpha, beta (left): \t", [bird.alpha_l, bird.beta_l])
        print("alpha, beta (right): \t", [bird.alpha_r, bird.beta_r])
        print("Fu, Fv, Fw: \t\t", bird.F)
        print("Tu, Tv, Tw: \t\t", bird.T)
        print("VFu, VFv, VFw: \t\t", bird.vortex_force_u, bird.vortex_force_v, bird.vortex_force_w)
        print("VTu, VTv, VTw: \t\t", bird.vortex_torque_u, bird.vortex_torque_v, bird.vortex_torque_w)
        print()

File: RK4Solver/cpp/flock.cpp
#include "types.hpp"
#include "utils.hpp"
#include "vortex.hpp"
#include "DiffEqs.hpp"
#include "bird.cpp"
#include "DiffEqs.cpp"
#include "vortex.cpp"

double start_time = time();
double tot_time = 0;

using Birds = std::vector<Bird>;
struct Flock{

    double h;
    double t;
    int N;
    int total_vortices;
    double total_dist;
    double max_r;
    bool LIA;
    int updates;
    int max_steps;
    double energy_punishment;
    double forward_reward;
    double crash_reward;
    Birds birds;
    BirdInits starting_conditions;
    Flock(  int N,
            double h,
            double t,
            BirdInits init_vals,
            bool LIA
         ){
        Flock & self = *this;
        self.h = h;
        self.t = t;
        self.N = N;
        self.LIA = LIA;

        self.total_vortices = 0.0;
        self.total_dist = 0.0;

        self.max_r = 1.0;
        self.updates = 0;
        self.max_steps = 100.0/self.h;

        self.energy_punishment = 2.0;
        self.forward_reward = 5.0;
        self.crash_reward = -100.0;

        for (size_t i : range(N)){
            self.birds.emplace_back(init_vals[i]);
        }
        self.starting_conditions = init_vals;
    }

    void reset(){
        Flock & self = *this;
        for (size_t i : range(self.N)){
            self.birds[i] = Bird(self.starting_conditions[i]);
        }
    }

    void update_bird(EnvAction action, int agent){
        Flock & self = *this;
        double thrust = action[0];
        Bird & bird = self.birds[agent];
        double start = time();
        self.update_angles(action, agent);
        tot_time += time() - start;
        Vorticies vortices = self.get_vortices(bird);
        bird.update(thrust, self.h, vortices);
        updates++;
        if (updates % 1000 == 0){
            //std::cout << tot_time/(time() - start_time) << std::endl;
            tot_time = 0;
            start_time = time();
            // std::cout << bird.VORTICES_LEFT.size() << std::endl;
        }
    }

    std::pair<bool, double> get_reward(EnvAction & action, int agent){
        Flock & self = *this;
        double reward = 0;
        bool done = false;
        Bird & bird = self.birds[agent];
        if (len(bird.X) >= 2 && bird.x > bird.X[len(bird.X)-2]){
            reward += self.forward_reward;
        }
        reward -= self.energy_punishment * action[0];

        if (self.crashed(bird)){
            done = true;
            reward = self.crash_reward;
        }

        if (bird.x > 500.0){
            done = true;
        }

        return std::make_pair(done, reward);
    }

    bool crashed(const Bird & bird){
        Flock & self = *this;
        if (bird.z <= 0 || bird.z >= 100){
            return true;
        }

        double lim = 2*PI;
        if (abs(bird.p) > lim || abs(bird.q) > lim ||  abs(bird.r) > lim){
            return true;
        }

        bool crash = false;
        for (int b : range(len(self.birds))){
            const Bird & other = self.birds[b];
            if (&other != &bird){
                double dist = (bird.xyz() - other.xyz()).norm();
                if (dist < bird.Xl/2.0){
                    crash = true;
                }
            }
        }
        return crash;
    }

    Vorticies get_vortices(Bird & curr){
        Flock & self = *this;
        Vorticies vortices;
        for (size_t b : range(len(self.birds))) {
            const Bird & bird = self.birds[b];
            if (&bird != &curr){
                for (const Vorticies & vorts : {bird.VORTICES_LEFT, bird.VORTICES_RIGHT}){
                    size_t i = 0;
                    //want first vortex ahead of current vortex
                    while (i < len(vorts) && vorts[i].x() < curr.x){
                        i = i+1;
                    }
                    const Vortex & v = vorts[i];
                    if (i < len(vorts) && v.x() >= curr.x){
                        // TODO: what about the x dimention? Don't you care about the radius in all 3 dimentions? Although I don't supos it matters that much
                        double r = sqrt(sqr(curr.y - v.y()) + sqr(curr.z - v.z()));
                        if (r < self.max_r){
                            vortices.push_back(v);
                        }
                    }
                }
            }
        }
        return vortices;
    }

    Observation get_observation(int agent){
        Flock & self = *this;
        Vector3d force = self.birds[agent].F;
        Vector3d torque = self.birds[agent].T;

        Bird & bird = self.birds[agent];
        Observation obs;
        extend(obs, force);
        extend(obs, torque);
        extend(obs, {bird.x, bird.y, bird.z});
        extend(obs, {bird.u, bird.v, bird.w});
        extend(obs, {bird.p, bird.q, bird.r});
        extend(obs, {bird.phi, bird.theta, bird.psi});
        extend(obs, {bird.alpha_l, bird.beta_l, bird.alpha_r, bird.beta_r});
        std::vector<Bird *> nearest = bird.seven_nearest(self.birds);
        for (Bird * otherp : nearest){
	        Bird & other = *otherp;
            extend(obs, {other.x - bird.x, other.y - bird.y, other.z - bird.z});
            extend(obs, {other.u - bird.u, other.v - bird.v, other.w - bird.w});
        }
        return obs;
    }

    void update_vortices(double vortex_update_frequency){
        Flock & self = *this;
        for (size_t b : range(len(self.birds))){
            Bird & bird = self.birds[b];
            bird.shed_vortices();

            if (self.LIA){
                bird.update_vortex_positions(bird.VORTICES_RIGHT, self.h*vortex_update_frequency);
                bird.update_vortex_positions(bird.VORTICES_LEFT, self.h*vortex_update_frequency);
            }

            //remove expired vortices
            if (len(bird.VORTICES_RIGHT) > 1.0/(self.h*vortex_update_frequency)){
                pop0(bird.VORTICES_LEFT);
                pop0(bird.VORTICES_RIGHT);
            }
        }
    }

    void update_angles(EnvAction action,int agent){
        Flock & self = *this;
        Bird & bird = self.birds[agent];
        double limit_alpha = PI/6.0;
        double limit_beta = PI/4.0;
        double new_al = bird.alpha_l + action[1];
        if (new_al > limit_alpha)
            new_al = limit_alpha;
        if (new_al < -limit_alpha)
            new_al = -limit_alpha;
        bird.alpha_l = new_al;

        double new_bl = bird.beta_l + action[2];
        if (new_bl > limit_beta)
            new_bl = limit_beta;
        if (new_bl < -limit_beta)
            new_bl = -limit_beta;
        bird.beta_l = new_bl;

        double new_ar = bird.alpha_r + action[3];
        if (new_ar > limit_alpha)
            new_ar = limit_alpha;
        if (new_ar < -limit_alpha)
            new_ar = -limit_alpha;
        bird.alpha_r = new_ar;

        double new_br = bird.beta_r + action[4];
        if (new_br > limit_beta)
            new_br = limit_beta;
        if (new_br < -limit_beta)
            new_br = -limit_beta;
        bird.beta_r = new_br;
    }


    Birds get_birds(){
        return this->birds;
    }
};
